org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/typehierarchy/HierarchyInformationControl.java

```diff
@@ -138,9 +138,23 @@ protected TreeViewer createTreeViewer(Composite parent, int style) {
 
 			@Override
 			protected void inputChanged(Object input, Object oldInput) {
-				visited= new HashSet<>();
-				super.inputChanged(input, oldInput);
-				visited= null;
+				runWithVisited(() -> super.inputChanged(input, oldInput));
+			}
+
+
+			@Override
+			public void expandToLevel(Object elementOrTreePath, int level, boolean disableRedraw) {
+				runWithVisited(
+						() -> super.expandToLevel(elementOrTreePath, level, disableRedraw));
+			}
+
+			private void runWithVisited(Runnable r) {
+				try {
+					visited= new HashSet<>();
```

Comment:
i think this should check the requirement, that visited==null to prevent against wrong nested calls